### PR TITLE
Feature/as 565 cache profile loading in service

### PR DESCRIPTION
**Pull request by lvnhmd**

### CODE REVIEW

File: lib/auth/profile.js [hunk 1]
```diff
@@ -8,13 +9,33 @@ module.exports = (endpoint) => {
...
-    return request(`/me`, { headers })
-      .then(response => response.json.data)
-      .catch(() => null);
+  return (token, session) => {
+
+    return Promise.resolve()
+      .then(() => {
+        if (!session.profile ||
+          (session.profile &&
+            moment(new Date()).isAfter(moment(session.profile.expiresAt).format()))) {
```
Comment (lennym) [inline lib/auth/profile.js]: The `new Date()` argument in the calls to moment can be removed. `moment()` will default to the current date/time if no argument is passed.

File: lib/auth/profile.js [hunk 2]
```diff
@@ -8,13 +9,33 @@ module.exports = (endpoint) => {
...
+
+          const headers = {
+            Authorization: `bearer ${token}`
+          };
+
+          return request(`/me`, { headers })
+            .then(response => {
+              const p = response.json.data;
+              p.expiresAt = moment.utc(moment(new Date()).add(600, 'seconds')).valueOf();
+              return p;
```
Comment (joefitter) [inline lib/auth/profile.js]: I would write this as:
```js
.then(({ json: { data } }) => ({
  ...data,
  expiresAt: moment.utc(moment().add(600, 'seconds')).valueOf()
}))
```
but Lenny would probably slap me

Comment (joefitter) [reply lib/auth/profile.js]: *_however_* As per Lenny's comment, the `new Date()` bit isn't needed here either

File: test/unit/profile/index.spec.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        expect(session.profile.lastName).toEqual('Last');
+      });
+  });
+
+  it('should not call profile api if a profile exists in the session', () => {
+    const session = {
+      profile: {
+        firstName: 'Someone',
+        lastName: 'Else',
+        expiresAt: moment.utc(moment(new Date()).add(60, 'seconds')).valueOf()
```
Comment (joefitter) [inline test/unit/profile/index.spec.js]: `new Date` not needed here

File: test/unit/profile/index.spec.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        expect(session.profile.lastName).toEqual('Else');
+      });
+  });
+
+  it('should call profile api if a stale profile exists in the session', () => {
+    const session = {
+      profile: {
+        firstName: 'Someone',
+        lastName: 'Else',
+        expiresAt: moment.utc(moment(new Date()).subtract(660, 'seconds')).valueOf()
```
Comment (joefitter) [inline test/unit/profile/index.spec.js]: `-- new Date()`

File: lib/auth/profile.js (hunk 2 above)
Comment (lennym) [reply lib/auth/profile.js]: > but Lenny would probably slap me

@joefitter Correct. I don't mind the `...` but that destructuring... 😱

Comment (joefitter) [reply lib/auth/profile.js]: FINE, let's merge it. I'll fix this code some time in the future. destructure all the things

**Pull request merged**